hail/src/main/scala/is/hail/types/physical/stypes/concrete/SInsertFieldsStruct.scala

```diff
@@ -0,0 +1,159 @@
+package is.hail.types.physical.stypes.concrete
+
+import is.hail.annotations.Region
+import is.hail.asm4s.{Code, Settable, TypeInfo, Value}
+import is.hail.expr.ir.{EmitCode, EmitCodeBuilder, EmitSettable, IEmitCode}
+import is.hail.types.physical.stypes.interfaces.{SBaseStruct, SBaseStructCode, SStructSettable}
+import is.hail.types.physical.stypes.{EmitType, SCode, SType}
+import is.hail.types.physical.{PCanonicalStruct, PType}
+import is.hail.types.virtual.{TStruct, Type}
+import is.hail.utils.BoxedArrayBuilder
+
+case class SInsertFieldsStruct(virtualType: TStruct, parent: SBaseStruct, insertedFields: IndexedSeq[(String, EmitType)]) extends SBaseStruct {
+  override def size: Int = virtualType.size
+
+  lazy val insertedFieldMap: Map[String, Int] = insertedFields.map(_._1).zipWithIndex.toMap
+
+  def getFieldIndexInNewOrParent(idx: Int): Either[Int, Int] = {
+    val fieldName = virtualType.fieldNames(idx)
+    insertedFieldMap.get(fieldName) match {
```

Comment:
Seems a bit unfortunate to have to go through field names here. What if instead of caching the `insertedFieldMap`, you save `fieldInBase: IndexedSeq[Boolean]` and `virtualToPhysicalIndex: IndexedSeq[Int]`, which maps field index to either `parent` or `insertedFields` index? (Or even store parent indices as negative ints.)

Reply:
Good call, fixed. Still need to use strings to pass through to parent.